\documentclass[journal,transmag]{IEEEtran}

% ..........................................................................
% Packages, configuration settings, and macro definitions.

\usepackage[pdftex]{graphicx}
\graphicspath{figures}
\DeclareGraphicsExtensions{.pdf,.jpeg,.png}

\usepackage[pdftex,rgb,dvipsnames,svgnames,hyperref,table]{xcolor}

\usepackage[pdftex,breaklinks=true,colorlinks=true,
    bookmarks=false,pdfhighlight=/O,
    urlcolor=DarkBlue,citecolor=DarkRed,linkcolor=DarkBlue]{hyperref}

\usepackage[cmex10]{amsmath}
\interdisplaylinepenalty=2500

\usepackage{amssymb}
\usepackage{amsfonts}
\usepackage{multicol}
\usepackage{multirow}
\usepackage{enumitem}
\usepackage{accsupp}
\usepackage{array}
\usepackage[caption=false,font=footnotesize]{subfig}
\usepackage{booktabs}
\usepackage{xspace}
\usepackage{soul}
\usepackage{url}
\usepackage{hyphenat}
\usepackage[english]{babel}
\usepackage[square,numbers,sort&compress]{natbib}

% Correct bad hyphenation here
\hyphenation{}

\newcommand{\email}[1]{\href{mailto:#1}{#1}}

\usepackage{pdfcomment}
\newcommand{\comment}[3]{\pdfmarkupcomment[markup=Highlight,color=yellow,author={#2}]{#1}{#3}}

\usepackage{tabularx}
% ..........................................................................
% Body.

\begin{document}

\markboth{IEEE Transactions on Biomedical Engineering}{Standards for whole-cell modeling}

\title{Toward standards for tomorrow's whole-cell models}

\author{
    Dagmar Waltemath$^*$,
    Falk Schreiber, 
    Jonathan R. Karr, 
    Chris J. Myers,~\IEEEmembership{Senior~Member,~IEEE,} \\
    Michael Hucka,
    Frank T. Bergmann,
    Vijayalakshmi Chelliah,
%   Marcus Krantz,
    Wolfram Liebermeister,
%   Pedro Mendes,
    %   Pinar Pir,
    %%%%%%%%%%%%%%%%%%%
    Begum Alaybeyoglu,\\
    Arne T. Bittig,
    Paulo E. Pinto Burke, 
    Yin Hoon Chew,
    Rafael S. Costa,
    Joseph Cursons, 
    Tobias Czauderna,\\
    Muhammad Haseeb,
    Denis Kazakiewic, 
    Ilya Kiselev,
    Vincent Knight-Schrijver,
    Christian Kn\"{u}pfer,
    Matthias K\"{o}nig,\\
    Nikita Mandrik,
    J. Kyle Medley,
    Sucheendra K. Palaniappan,
    Martin Scharm,
    Mahesh Sharma,
    Kieran Smallbone,\\
    Je-Hoon Song,
    Tom Theile,
    Namrata Tomar,
    Jannis Uhlendorf,
    Markus Wolfien,\\
    James T. Yurkovich,
    Yan Zhu, and
    Anna Zhukova
    \thanks{
        Manuscript received XXX XX, 2015; revised XXX XX, 201X; accepted XXX XX, 201X. Date of publication XXX XX, 201X; date of current version XXX XX, 201X.
        This work was supported in part by the Volkswagen Foundation (Grant to D. W. and F. S.) and the James S. McDonnell Foundation (Postdoctoral Fellowship Award in Studying Complex Systems to J. R. K.).
        \textit{Asterisk indicates corresponding author.}
    }
    \thanks{D. Waltemath, A. T. Bittig, M. Scharm, T. Theile, and M. Wolfien are with the Institute of Computer Science, University of Rostock, 18051 Rostock, Germany (e-mail: \email{dagmar.waltemath@uni-rostock.de}; \email{arne.bittig@uni-rostock.de}; \email{martin.scharm@uni-rostock.de}; \email{tom.theile@uni-rostock.de}; \email{markus.wolfien@uni-rostock.de}).}
    \thanks{F. Schreiber and T. Czauderna are with the Faculty of Information Technology, Monash University, Clayton, VIC 3800, Australia and also with the Institute of Computer Science, Martin Luther University Halle-Wittenberg, 06108 Halle, Germany (e-mail: \email{falk.schreiber@monash.edu}; \email{tobias.czauderna@monash.edu}).}
    \thanks{J.~R. Karr is with the Department of Genetics \& Genomic Sciences, Icahn School of Medicine at Mount Sinai, New York, NY 10029 USA (e-mail: \email{karr@mssm.edu}).}
    \thanks{C.~J. Myers is with the Department of Electrical and Computer Engineering, University of Utah, Salt Lake City, Utah 84112, USA (e-mail: \email{myers@ece.utah.edu}).}
    \thanks{M. Hucka is with the Department of Computing and Mathematical Sciences, California Institute of Technology, Pasadena, CA 91125, USA (e-mail: \email{mhucka@caltech.edu}).}
    \thanks{F.~T. Bergmann is with BioQuant, University of Heidelberg, 69120 Heidelberg, Germany (e-mail: \email{fbergman@caltech.edu}).} 
    \thanks{V. Chelliah is with the European Bioinformatics Institute, European Molecular Biology Laboratory, Cambridge CB10 1SD, UK (e-mail: \email{viji@ebi.ac.uk}).}  
    \thanks{M. Krantz and J. Uhlendorf are with the Department of Biology, Humboldt University of Berlin, 10115 Berlin, Germany (e-mail: \email{marcus.krantz@biologie.hu-berlin.de}; \email{jannis.uhlendorf@hu-berlin.de}).}
    \thanks{W. Liebermeister and M. K\"{o}nig are with the Institute of Biochemistry, University Medicine Charit\'{e} Berlin, 10117 Berlin, Germany (e-mail: \email{wolfram.liebermeister@gmail.com}; \email{matthias.koenig@charite.de}).}
    %\thanks{P. Mendes is with the Manchester Institute of Biotechnology and the School of Computer Science, University of Manchester, Manchester M1 7DN, UK and also with the Center for Quantitative Medicine and the Department of Cell Biology, University of Connecticut Health Center, Farmington, CT 06030, USA (e-mail: \email{pedro.mendes@manchester.ac.uk}).}
    \thanks{P. Pir and V. Knight-Schrijver are with the Babraham Institute, Cambridge CB22 3AT, UK (e-mail: \email{pinar.pir@babraham.ac.uk}).}
    \thanks{B. Alaybeyoglu is with the Department of Chemical Engineering, Bo\v{g}azi\c{c}i University, Bebek 34342, Turkey (e-mail: \email{begum.alaybeyoglu@boun.edu.tr}).}
    \thanks{P.~E. Pinto Burke is with the Institute of Science and Technology, Federal University of São Paulo, Brazil.}
    \thanks{Y.~H. Chew is with the Centre for Synthetic and Systems Biology, University of Edinburgh, Edinburgh EH9 3BF, UK (e-mail: \email{yinhoon.chew@ed.ac.uk}).}
    \thanks{R.~S. Costa is with the Centre of Intelligent Systems-IDMEC, Instituto Superior Técnico, University of Lisbon, 1049-001 Lisboa, Portugal (e-mail: \email{rafael.s.costa@tecnico.ulisboa.pt}).}
    \thanks{J. Cursons is with the Department of Biomedical Engineering, School of Engineering, University of Melbourne, Parkville, VIC 3010, Australia (e-mail: \email{joseph.cursons@unimelb.edu.au}).}
    \thanks{M. Haseeb is with the Department of Bioinformatics, Mohammad Ali Jinnah University, Islamabad, Pakistan.}
    \thanks{D. Kazakiewic is with the Center for Statistics, Universiteit Hasselt, Hasselt BE3500, Belgium, and also with the Center for Innovative Research, Medical University of Białystok, Białystok 15-089, Poland (e-mail: \email{dzianis.kazakevich@uhasselt.be}).}
    \thanks{I. Kiselev is with the Design Technological Institute of Digital Techniques, Siberian Branch of the Russian Academy of Sciences, Novosibirsk 630090, Russia.}
    \thanks{C. Kn\"{u}pfer is with the Institut f\"ur Informatik, University of Jena, 07743 Jena, Germany (e-mail: \email{christian.knuepfer@uni-jena.de}).}
    \thanks{N. Mandrik is with the Sobolev Institute of Mathematics, Siberian Branch of the Russian Academy of Sciences, Novosibirsk 630090, Russia (e-mail: \email{manikitos@gmail.com}).}
    \thanks{J.~K. Medley is with the Department of Bioengineering, University of Washington, Seattle, WA 98195, USA (e-mail: \email{medjk@comcast.net}).}
    \thanks{S.~K. Palaniappan is with the Rennes - Bretagne Atlantique Research Centre, Institute for Research in Computer Science and Automation, 35042 Rennes Cedex, France.}
    \thanks{M. Sharma is with the Department of Pharmacoinformatics, National Institute of Pharmaceutical Education and Research, Punjab 160062, India.}
    \thanks{K. Smallbone is with the Manchester Centre for Integrative Systems Biology, University of Manchester, Manchester M1 7DN, UK (e-mail: \email{kieran.smallbone@manchester.ac.uk}).}
    \thanks{J.-H. Song is with the Department of Bio and Brain Engineering, Korea Advanced Institute of Science and Technology, Daejeon 305-701, Republic of Korea (e-mail: \email{song.je-hoon@kaist.ac.kr}).}
    \thanks{N. Tomar is with the Department of Dermatology, University Medicine, Friedrich-Alexander University of Erlangen-N\"urnberg, 91052 Erlangen, Germany.}
    \thanks{J. T. Yurkovich is with the Department of Bioengineering, University of California, San Diego, La Jolla, CA 92093, USA (e-mail: \email{jyurkovich@ucsd.edu}).}
    \thanks{Y. Zhu is with Monash Institute of Pharmaceutical Sciences, Monash University, Parkville, VIC 3052, Australia.}
    \thanks{A. Zhukova is with the Institut de Biochimie et Génétique Cellulaires, National Center for Scientific Research, and also with the University of Bordeaux, France, 33077 Bordeaux Cedex, France (e-mail: \email{zhutchok@gmail.com}).}
    \thanks{Digital Object Identifier 10.1109/TBME.XXXX.XXXXXXX}
}

\maketitle

\begin{abstract}
Whole-cell models are promising tools for biological research, bioengineering, and medicine. 
However, substantial work remains before complete and accurate whole-cell models can be created in a consistently reproducible form.
Among the advances needed are a strong theoretical understanding of multi-algorithm modeling, a standardized whole-cell modeling language, and an efficient general-purpose simulator.
We organized the 2015 Whole-Cell Modeling Summer School to teach whole-cell modeling as well as to evaluate the need for new modeling standards by attempting to encode a recently published whole-cell model in SBML.
We propose three SBML extensions to support transparent, reproducible whole-cell modeling: support for multi-algorithm models, support for particle-based state representation, and support for template reactions. In addition, we describe several new software tools and databases that are needed to enable researchers to encode and simulate whole-cell models, including a user-friendly graphical model editor and a parallelized simulator. We also propose several new extensions to SGBN.
Together these new standards and software tools would accelerate whole-cell modeling.
\end{abstract}

\begin{IEEEkeywords}
Whole-cell modeling, Systems biology, Computational biology, Mathematical modeling, Simulation, Standards, Education
\end{IEEEkeywords}

\IEEEpeerreviewmaketitle

\section{Introduction}

\IEEEPARstart{O}{ver} the past twenty years, computational modeling has become an essential and powerful tool for biological research, bioengineering, and medicine to analyze high-throughput molecular measurements and understand the molecular details of complex biological systems. Computational modeling has  been used to identify new metabolic genes~\cite{Reed2006}, to add metabolic pathways to bacteria~\cite{Lee2012}, and to identify potential new antimicrobial drug targets~\cite{Lee2009}.
Computational models also have the potential to enable bioengineers to design new microorganisms for industrial applications such as chemical synthesis, biofuel production, and waste decontamination, as well as to enable clinicians to tailor therapy to individual patients. Realizing this potential requires more comprehensive and accurate computational models that are capable of predicting cellular behavior from genotype. Realizing this potential also requires improved simulation tools, as well as standardized tools for storing and exchanging models, simulation experiments, and model visualizations~\cite{Macklin2014,Karr2015,Karr2015b,hucka2015promoting,Klipp07,path2models2013}.

Recently, researchers at Stanford University developed the first whole-cell model of the gram-positive bacterium \textit{Mycoplasma genitalium}~\cite{Karr2012}. The model represents the life cycle of a single Mycoplasma cell including the copy number dynamics of each metabolite, RNA, and protein species and accounts for every known gene function. The model is composed of 28 submodels, each of which is implemented using different mathematical formalisms including ordinary differential equations (\emph{ODE}s), flux balance analysis (\emph{FBA}), and Boolean rules (\emph{BR}s), and trained using different experimental data.

The \textit{M. genitalium} whole-cell model was implemented in MATLAB, is available open-source under the MIT license~\cite{wholeCell}, and is extensively documented. This has enabled others to use the model for their own research. 

However, the \textit{M. genitalium} whole-cell model simulation software is not transparent or reusable. The \textit{M. genitalium} whole-cell model simulation software is also not user-friendly, computationally efficient, or easily maintainable. Covert and colleagues have developed several software programs that include WholeCellKB~\cite{Karr2013}, WholeCellSimDB~\cite{Karr2014}, and WholeCellViz~\cite{Lee2013} to provide user-friendly interfaces on top of their whole-cell simulation software. However, significant domain expertise is still required to use the \textit{M. genitalium} model or construct new whole-cell models. Therefore new modeling standards and simulation tools are needed to enable more researchers to develop and simulate their own whole-cell models. Such standards and software tools would enable researchers to develop whole-cell models more quickly, to explore models more deeply, and to evaluate models more rigorously. Furthermore, standards would simplify the submission process to model repositories such as BioModels~\cite{juty2015biomodels,chelliah2015biomodels}. This would make models more searchable, retrievable, reusable, and comparable as well.

Several software-agnostic standards in systems biology have already been developed by the COmputational Modeling in BIology NEtwork (\emph{COMBINE})~\cite{le2011meeting}, including the Systems Biology Markup Language (\emph{SBML})~\cite{hucka2003}, the Cell Markup Language (\emph{CellML})~\cite{hedley_2001b}, the Simulation Experiment Description Markup Language (\emph{SED-ML})~\cite{sedml2011}, and the Systems Biology Graphical Notation (\emph{SBGN})~\cite{LeNovereHMMSS09}. SBML and CellML are formats for representing mathematical models employing common formalisms such as ODE, logical, and FBA models. Both have been used to build thousands of models of various intracellular pathways. SED-ML is a language for describing computational experiments. SED-ML enables scientists to reproduce a simulations by completely describing simulation setups, including the simulation algorithm and every parameter value. SBGN includes three languages for creating visual representations of models. None of these community standards have been used to construct, simulate, or visualize models as complex as the \textit{M. genitalium} whole-cell model.

We organized the 2015 Whole-Cell Modeling Summer School to train students in whole-cell modeling, as well as to evaluate the need for new standards for whole-cell modeling. The majority of the school was focused on trying to encode the \textit{M. genitalium} model using SBML; this served both as a method to train students as well as an exercise to evaluate SBML's potential for encoding whole-cell models. The ultimate scientific goal of the school was to develop an open-source whole-cell model encoded in SBML and simulated using SED-ML.

In what follows, we describe the summer school, outline our progress toward encoding the \textit{M. genitalium} model using SBML, and propose several SBML and SGBN extensions to support whole-cell modeling. We begin with a summary of the summer school, then describe our progress towards encoding the \textit{M. genitalium} model using SBML, and finally, describe the SBML and SBGN extensions and software tools found to be needed to support whole-cell modeling.

\section{The 2015 Whole-Cell Modeling Summer School}
We organized the summer school to teach students how to build and encode models using COMBINE standards by attempting to encode the \textit{M. genitalium} model using software-independent, standard representation formats.

\subsection{Organization}
The Whole-Cell Modeling Summer School was held March 9-13, 2015 at the University of Rostock in Rostock, Germany. The school was organized by Dagmar Waltemath and Falk Schreiber and supported by the Volkswagen Foundation. 45 students, nine instructors, and two organizers participated in the five-day school.

The school began with two lectures that introduced whole-cell modeling and the existing systems biology standards. Jonathan Karr from the Icahn School of Medicine at Mount Sinai, USA presented an overview of whole-cell and multi-algorithm modeling. Michael Hucka from the California Institute of Technology, USA presented an overview of the SBML, SED-ML, and SBGN standards; several open-source software tools that support these standards; and the COMBINE initiative. We also organized three discussions on multi-algorithm model composition, particle-based state representation, and random number generation. 

The majority of the school was dedicated to hands-on active learning sessions in which students learned about whole-cell modeling and the COMBINE standards by trying to encode parts of the \textit{M. genitalium} whole-cell model using SBML. Students were divided into ten groups of four to five students, each of which was challenged to encode one or more submodels using SBML. Each group was led by an experienced instructor.

Each day concluded with brief progress reports from each group. This facilitated discussion on common encoding challenges and model integration, and provided an opportunity for groups to obtain feedback from each other.

We also organized a poster session, as well as several evening social activities to provide the students opportunities to network with each other and the instructors.

\subsection{Educational outcomes}
\comment{We surveyed the students to assess the educational outcome of the school.}{Karr}{What did the survey actually show?}
Most students reported gaining deep knowledge of whole-cell modeling, increased appreciation for reproducible science, and increased understanding of the SBML, SED-ML, and SBGN standards. Many students also reported learning about open-source modeling software tools relevant to their own research.

In addition, many of the students reported that the school expanded their scientific network. Several students commented that the school introduced them to potential postdoctoral positions and next year's whole-cell modeling summer school (\href{http://www.wholecell.org/school-2016}{http://www.wholecell.org/school-2016}).

\subsection{Lessons learned for organizing research-based schools}
We learned several valuable lessons about how to best organize an open-ended, research-based school. First, we conclude that research-based schools should clearly outline the expected background knowledge and learning objectives and have well-organized learning activities. This helps students make informed decisions about whether to participate in the school, know how to prepare for the school, and learn efficiently.
Second, we conclude that students greatly enjoy learning through open research problems rather than through prescribed training exercises. This makes students feel engaged, challenged, and connected to research. This also helps students build practical skills to complement their foundational undergraduate and graduate training.
Third, we conclude that open-ended project-based schools require a high teacher-to-student ratio, a flexible schedule, and multidisciplinary project teams. A high teacher to student ratio allows students to get feedback and iterate through potential solutions quickly. A flexible schedule enables impromptu lectures and discussions. Multidisciplinary teams enable students to work through difficult problems by drawing on perspectives from multiple fields. 

\section{Toward an SBML-encoded whole-cell model}
In addition to training young computational systems biology researchers, the second goal of the school was to attempt to encode the \textit{M. genitalium} whole-cell model (\url{https://github.com/CovertLab/WholeCell/releases/tag/v1.1}) in SBML. To achieve this goal, most of the course was devoted to active learning sessions in which students were challenged to encode submodels of the \textit{M. genitalium} in SBML, integrate submodels into a single model, and simulate models using SED-ML. During these sessions, the students and instructors were divided into ten groups. Eight of the groups were tasked with encoding one or more submodels. The ninth group was tasked with developing a standards-compliant scheme to integrate the submodels into a single model. This group was responsible for defining the global state variables and submodel interfaces and developing a SED-ML scheme to simulate the integrated model. The tenth group was responsible for annotating the model and helping the other groups visualize their submodels. Table~SI lists the ten groups and all of the students and instructors. 

\subsection{submodel encoding}
The eight submodel encoding groups pursued various strategies to encode the submodels using SBML. Several of the groups encoded submodels by first reading the submodel documentation, then drawing pathway diagrams using software tools such as CellDesigner~\cite{funahashi2008celldesigner} and VANTED~\cite{Rohn2012}, and finally writing scripts to generate SBML models from their diagrams using libSBML~\cite{bornstein_2008}. Other groups used modeling software tools such as Antimony~\cite{Smith2009}, BioUML~\cite{Kolpakov2006}, COPASI~\cite{Mendes2009}, and iBioSim~\cite{Madsen2012} to encode submodels based on their documentation. The metabolism and transcription groups also used simulation libraries such as libRoadRunner~\cite{Somogyi17062015} and COBRApy~\cite{ebrahim2013cobrapy}. A few of the groups encoded submodels by converting the MATLAB code to SBML. These groups then generated SBGN diagrams from their SBML to better understand their submodels.

The groups encountered several challenges to encoding the \textit{M. genitalium} submodels in SBML. First, most of the groups had to spend a significant amount of time reading the MATLAB code and documentation to understand the details of the \textit{M. genitalium} submodels because the connection between the submodels and the associated pathway/genome database is not transparent, many of the submodels details are implemented directly in MATLAB code rather than in a transparent language such as SBML, and the documentation only provides overviews of the submodels. Fortunately, one of the authors of the \textit{M. genitalium} model was available to answer questions about the model.

A second challenge to encoding the submodels in SBML was to encode serially executed MATLAB submodels, because SBML is not a programming language and thus does not offer control over the order of simulation execution. This fundamental difference between programming languages and SBML makes direct quantitative reproduction of the \textit{M. genitalium} model impossible. Most of the groups decided to tackle this problem by formalizing MATLAB submodels as discrete stochastic models and simulating them using the stochastic simulation algorithm of Gillespie~\cite{gillespie1977} or other approximate algorithms. For several of the submodels, this conversion imposed an explicit internal submodel timescale that was not present in the original MATLAB submodel due to the lack of kinetic data for the corresponding pathway.

The fact that SBML is not a programming language and does not expose methods for arbitrary random number generation also made it a challenge for groups to encode the random algorithms used by the MATLAB submodels in SBML. For example, the MATLAB translation submodel includes a random algorithm that assigns amino acids to individual polypeptides. Because this algorithm is not equivalent to the Gillespie algorithm, and plain SBML does not provide random number functions, the algorithm cannot easily be encoded in SBML. Most of the groups also solved this problem by formalizing submodels as stochastic models. Even if it were possible to transcode the MATLAB submodels directly into SBML form, it would still be difficult to quantitatively reproduce the MATLAB simulations because SBML does not expose control over the random number generator algorithm or seed. Consequently, it would only be feasible to compare the first two moments of the MATLAB and transcoded model simulations.

To encode many of the submodels in SBML, the groups also had to either enumerate the hybrid population/particle-based state representation used by the MATLAB submodels, or approximate the MATLAB submodels. The groups responsible for the transcription and translation submodels chose to approximate the MATLAB submodels by eliminating the internal dynamics of the polymerization of each RNA and polypeptide. Consequently, these submodels no longer track the progress of individual RNA polymerases and ribosomes, account for base-specific transcription or translation rates, or predict RNA polymerase collisions. The groups responsible for the DNA submodels including replication, replication initiation, and transcriptional regulation, chose to enumerate the sparse chromosome representation used by the MATLAB model by creating Boolean indicator variables to represent the existence and protein-binding status of each chromosome base. This enumerated representation requires millions of variables. Consequently, the corresponding SBML XML files are impractical to read and computationally expensive to simulate. Enumerating the rules that govern the joint values of the enumerated variables, such as the rules that represent the steric effects of DNA-bound proteins by preventing proteins from binding neighboring bases, is also impractical. Furthermore, the enumerated SBML files are impractical for humans to read, edit, or maintain.

The lack of universal SBML simulator support for arrays was another challenge to encoding the MATLAB submodels in SBML. All of the groups overcame the lack of array support by enumerating individual array elements of the MATLAB and all matrix algebra computations. This creates verbose SBML files that are more difficult to interpret, maintain, and edit. Enumerating the matrix algebra computations also increases the computational cost of simulation.

Together, these five challenges made it very difficult for the groups to encode most of the MATLAB submodels in SBML. Going forward, SBML and the SBML simulators should be expanded to provide support for random number generation, particle-based state representation, and arrays.

\subsection{Model integration}
The integration group was responsible for assembling the individual submodels into a single model. This included devising a scheme for representing global state variables, defining the interfaces exposed by the submodels to the global state variables, and developing a method to manage  multiple submodels writing shared state variables simultaneously. The integration group defined the global state variables as the union of all state variables shared by at least two submodels rather than by explicitly defining a set of global state variables as done by the original MATLAB-based simulation. The advantages of this approach are that submodel developers are not also required to develop global state variables and that it minimizes the number of global state variables. The disadvantages of this approach are that the total set of variables is less transparent and that it requires users to learn all of the submodels and their naming conventions to analyze model simulations.

The integration group standardized the interfaces exposed by the individual submodels by defining a variable naming convention. This naming convention ensures, for example, that the copy numbers of each protein species are represented by variables with the same names in each of the submodels. This convention makes it clear how multiple local submodel variables map onto the same global variable. Specifically, the integration group chose to use the same variable names as those used by the MATLAB implementation. Matrix and particle-based variables were enumerated by creating multiple variables with names containing additional suffixes to indicate their identity.

The primary challenge faced by the integration group was how to handle concurrent editing of shared state variables by multiple submodels. \comment{The integration group explored several potential strategies to manage concurrent writing.}{Karr}{Chris Myers, can you elaborate?} First, they explored sequentially simulating the submodels and updating the global state variables. This avoids the need for more complex strategies to merge variable changes. However, under this approach submodels are simulated with different variable values within each time step. Consequently, simulation predictions are sensitive to the submodel execution order. 

The integration group also explored several more complex submodel integration strategies that enable all of the submodels to be simulated with the same variable values within each time step. These strategies included reducing the submodel integration time step such that submodels do not request conflicting variable changes; dividing each of the shared state variables into separate, independent sub-variables for each submodel, simulating the submodels in parallel, and merging the sub-variables to compute the update global values; and using semaphores to manage concurrent variable changes whereby at each time step submodels request sets of atomic state variables changes and a controller decides which change sets are processed. Each of these strategies has different advantages and disadvantages. The first strategy is the simplest to understand and implement, but is computationally expensive. The second strategy is simple to implement and computationally efficient for independent variables, but is difficult to implement for sets of non-independent variables such as those that represent the protein occupancy of the chromosome. The third strategy is the most complex to implement, but is more general than the second strategy and more computationally efficient than the first. The integration group tested their submodel integration strategies using iBioSim because iBioSim is one of the only SBML simulators to support all of the existing needed SBML packages including Hierarchical Model Composition (``\emph{comp}''), Arrays, and Flux Balance Constraints (``\emph{fbc}'').

The lack of an SBML simulator that supports multi-algorithm model composition was another challenge to integrating the submodels. The integration group plans to overcome this limitation by adding support for multi-algorithm to iBioSim.

\subsection{Annotation, documentation, and visualization}
The documentation group was responsible for annotating the model.
The goal was to annotate every model entity with a cross reference to an external database such as ChEBI~\cite{Hastings2013} as in the case of small molecules and/or in terms of other model entities as in the case of chemical reactions.
The group wrote scripts to search molecular biology databases such as ChEBI for every entity contained in the \textit{M. genitalium} model.
The main problem faced by the documentation group was that many model entities are not currently represented by any molecular biology database.
This shortcoming can easily be overcome by proposing new ontology terms to the databases.
This problem highlights the need for expanded molecular biology databases to aggregate data on more biological molecules and interactions. 

The documentation group also helped the other groups visualize their submodels by providing advice on SBGN and SBGN-compatible diagramming tools such as VANTED~\cite{Rohn2012}.
The main visualization problem encountered by the documentation group was that whole-cell models require large diagrams that must be manually arranged to produce intuitive visualizations.

\subsection{Progress}
The students produced preliminary SBML and SBGN-ML versions of many of the \textit{M. genitalium} whole-cell model submodels. However, significant work remains to finish encoding the submodels, integrate the submodels into a single model, and test the submodels and the combined model. Complete drafts are only available for a few of the simplest submodels. The current drafts of most of the more complex submodels are also greatly simplified compared to their MATLAB versions. For example, the transcription and translation submodels drafts do not represent the polymerization of individual bases and the DNA submodels do not account for the protein occupancy of the chromosome. In addition, none of the SBML-encoded submodels have been thoroughly tested by replicating the unit tests from the MATLAB implementation. Furthermore, none of the SBML submodels have been thoroughly documented and complete SBGN diagrams are not yet available for any of the submodels.

\begin{figure*}[bt!]
\centering
\includegraphics[width=\textwidth]{figure1/figure1.pdf}
\caption{\label{fig:1} Whole-cell modeling pipeline. First, researchers will assemble experimental data into pathway/genome databases. Second, researchers will use pathway/genome databases to construct submodels. Third, researchers will use multi-algorithm simulators to conduct in silico experiments. Lastly, researchers will analyze in silico experiments to learn new biology.}
\end{figure*}

\subsection{Future steps}
We hope to finish encoding the \textit{M. genitalium} submodels in SBML and integrate the SBML-encoded submodels into a single model that can be simulated by open-source software tools such as COPASI, libRoadRunner, BioUML, and iBioSim. Several students and instructors have continued to keep working and meeting online. Several students and instructors also plan to participate in a second meeting in October 2015 that will be held at the University of Utah, USA, immediately prior to the 2015 COMBINE Forum.

Going forward, we plan to publish SBML-encoded versions of each of the \textit{M. genitalium} submodels to BioModels, along with SED-ML tests, SBGN diagrams, and textual documentations. This would make the submodels searchable, retrievable, and reusable by other scientists. We believe this would be a valuable community resource. It would demonstrate how to build a whole-cell model and enable other researchers to build upon the \textit{M. genitalium} submodels. Ultimately, SBML, SBGN, the SBML and SBGN software tools also need to be extended to facilitate whole-cell modeling.

\section{Toward SBML-, SED-ML-, and SBGN-based standards for whole-cell modeling}
Prior to the school, SBML, SED-ML, and SBGN had never been used to represent whole-cell or similarly complex models. Consequently, not surprisingly, the summer school revealed several limitations of SBML, SBGN, and the existing simulation software for large models. Importantly, the summer school facilitated discussions among modelers, tool developers, and model curators about how to best expand the existing standards and software tools to overcome these limitations.

\subsection{Current limitations}
As discussed above, SBML does not support multi-algorithm modeling, particle-based state representation, or arbitrary random number generation and the current SBML simulators have limited support for arrays. Consequently, SBML cannot yet be used to efficiently represent models with large, combinatorial state spaces; represent arbitrary stochastic models; or arbitrary mathematical models involving matrix algebra operations. State spaces must be enumerated, stochastic models must be described using a ratio time scale and simulated using the Gillespie or other approximate algorithm, and matrix algebra operations must be expanded. For the \textit{M. genitalium} whole-cell model this would result in large, unmanageable XML files and millions of variables.

Furthermore, no general-purpose SBML simulation software is currently capable of efficiently simulating models involving millions of variables and no SBML simulation software supports multi-algorithm modeling. In addition, no SBML editing tool provides researchers a user-friendly interface for editing SBML files containing millions of variables. Consequently, currently, whole-cell model SBML files must be generated by scripts and cannot be easily edited.

The summer school also revealed several limitations of the graphical representation scheme SBGN and the existing SBGN viewers for whole-cell models. First, whole-cell modeling requires hybrid SBGN diagrams that contain all three types of nodes: Process Descriptions, Entity Relationships, and Activity Flows. Currently, SBGN diagrams can only contain one type of node. Second, whole-cell modeling requires new automatic layout algorithms suitable for large visualizations. The existing layout algorithms are unable to construct intuitive diagrams of many of the pathways that need to be represented in whole-cell models. Third, in order to effectively visualize complex SGBN diagrams of whole-cell models, SBGN viewers must be able to display diagrams at different levels of granularity by automatically reducing diagrams.

The summer school did not reveal any limitations of SED-ML for describing whole-cell model simulations.

\subsection{Standard extensions}
Taken together, expanded standards, simulation software tools, and databases are needed to facilitate accurate, reproducible whole-cell modeling. We propose three SBML expansions to facilitate whole-cell modeling. First, the SBML \emph{comp} package must be expanded to support models composed of submodels implemented with different simulation algorithms. Currently, the comp package only supports models composed of submodels each implemented using the same simulation algorithm. In addition, the existing SBML simulators must be expanded to support multi-algorithm simulations and/or new simulators must be developed that support the expanded package. This requires significantly more research to determine the best ways to integrate heterogeneous submodels, including rigorously evaluating the advantages and disadvantages of each of the schemes proposed by the integration group. Significant effort will also be needed to develop a parallelized simulator that is capable of quickly simulating complex whole-cell models.

Second, a new SBML package must be created to support hybrid population/particle-based state representations such as those used by BioNetGen \cite{Hlavacek2006, Hogg2014} and NFSim \cite{Sneddon2011}. In parallel, the existing SBML simulators must be expanded to support this new package and/or new SBML simulators must be developed. This would enable modelers to compactly describe and efficiently simulate models with large, combinatorial state spaces. The compact descriptions enabled by this package would also make models more transparent and easier to maintain and expand. 

A new SBML package must also be created to support reaction templates so that, for example, translation could be described using a single reaction template and arrays of mRNA-specific translation initiation rates and codon-specific elongation rates. Such reaction templates would enable whole-cell and other large models to be compactly described, and consequently easily interpretable, maintainable, and editable. By separating the mathematical descriptions and quantitative parameter values, reaction templates would also make the connection between dynamical models and the experimental data used to inform their parameter values more transparent. The new reaction templates could be expanded for backward compatibility with older SBML simulators.

New user-friendly graphical editors must also be developed to enable researchers to easily build SBML files that take advantage of these new features. These graphics editors must also allow researchers to transparently map model parameters onto experimental data organized in pathway/genome databases.

In addition, our molecular biology databases such as ChEBI must be expanded to enable researchers to rigorously define whole-cell models in terms of external entities. SBGN must also be expanded to support hybrid diagrams and the SBGN viewers must be expanded to support more automatic layout algorithms, automatic model reduction, and contextual zooming. These features would enable researchers to use SBGN to visualize whole-cell and other large models.

Together, these SBML, SBGN, software, and database expansions would enable more researchers to more easily build, manage, simulate, and reproduce whole-cell models and simulations. These new standards and software tools would also enable researchers to build more comprehensive and more accurate models, including of human cells. Ultimately, these new standards and software would enable whole-cell modeling to support rational biological design and personalized medicine.

\subsection{The whole-cell modeling pipeline}
We anticipate that such expanded standards and tools will enable a four step approach to whole-cell model-driven discovery (Fig.~\ref{fig:1}). First, researchers will assemble experimental data from numerous sources including databases such as SABIO-RK~\cite{Wittig2012} and UniProt~\cite{UniProt2015} into pathway/genome databases using software tools such as Pathway Tools~\cite{Karp2010} and WholeCellKB. Second, researchers will use pathway/genome databases and graphical modeling tools such as BioUML, COPASI, and iBioSim to build submodels and encode them using transparent languages such as SBML. Third, multi-algorithm simulators will be used to conduct in silico experiments. Lastly, software tools such as WholeCellSimDB and WholeCellViz will be used to discover new biology through exploring, visualizing, and analyzing in silico experiments.

\section{Conclusion}
The 2016 Whole-Cell Modeling Summer School provided 45 young scientists with hands-on training in whole-cell and multi-algorithm modeling by attempting to encode the \textit{M. genitalium} whole-cell model in SBML. Additional summer schools and courses are needed to provide students with deeper theoretical training in dynamical modeling, multi-algorithm modeling, model reduction, and parameter estimation, as well as practical training in model construction including data curation, model building, numerical optimization, model testing, and model analysis.

The summer school also made significant strides toward encoding the \textit{M. genitalium} whole-cell model using SBML for simulation by open-source software. The students developed preliminary SBML versions of all of the submodels of the \textit{M. genitalium} model. Since the summer school, several students have continued to encode the \textit{M. genitalium} model, and several of the students and instructors are participating in a second meeting prior to 2015 COMBINE Forum at the University of Utah in Salt Lake City, USA. Ultimately, we hope to publish an SBML-encoded version of the model to BioModels. 

Despite our progress, significant work remains to complete, test, integrate, simulate, and document the SBML-encoded version of the \textit{M. genitalium} model. The current version of SBML and its extensions cannot yet represent whole-cell models effectively; there is no simulation software program that could efficiently simulate an SBML-encoded whole-cell model; and there is no graphical editor that is capable of constructing, editing, or visualizing an SBML-encoded whole-cell model. SBML must be expanded to support multi-algorithm modeling, template reactions, particle-based state representation, and arrays, and in addition, an efficient simulation software system and a user-friendly model editor must be developed to enable modelers to easily construct and simulate whole-cell models. New parameter estimation, model testing, and visual analysis tools must also be developed to enable researchers to effectively use SBML-encoded whole-cell models for scientific research. In addition, our molecular biology databases must be expanded to facilitate whole-cell model annotation. Furthermore, SBGN and the SBGN viewers must also be expanded to support hybrid diagrams, automatic graph layout, automatic graph reduction, and contextual zooming.

In summary, we believe that whole-cell modeling has the potential to be an important tool for biological discovery, bioengineering, and medicine. Achieving this potential requires improved standards for describing whole-cell models, as well as new general-purpose simulation software for reproducibly simulating whole-cell models. In turn, this requires expanding the whole-cell modeling field including training additional young researchers.

\ifCLASSOPTIONcaptionsoff
  \newpage
\fi

\bibliographystyle{IEEEtran}
\bibliography{IEEEabrv,report}

% biography section
% 
% If you have an EPS/PDF photo (graphicx package needed) extra braces are needed around the contents of the optional argument to biography to prevent
% the LaTeX parser from getting confused when it sees the complicated \includegraphics command within an optional argument. (You could create
% your own custom macro containing the \includegraphics command to make things simpler here.)

\begin{IEEEbiography}[{\includegraphics[width=3.45in,keepaspectratio]{photos/group.jpg}}]{}
~\\
~\\
~\\
~\\
~\\
~\\
~\\
~\\
~\\
~\\
~\\
\textbf{2015 Whole-Cell Modeling Summer School} included the 56 researchers listed in Table~SI.
\end{IEEEbiography}


% You can push biographies down or up by placing a \vfill before or after them. The appropriate
% use of \vfill depends on what kind of text is on the last page and whether or not the columns are being equalized.

\vfill

% Can be used to pull up biographies so that the bottom of the last one is flush with the other column.
%\enlargethispage{-5in}

% ..........................................................................
% End.

\clearpage
\setcounter{table}{0}
\renewcommand{\thetable}{S\Roman{table}}

\begin{table*}[ht!]
\caption{2015 Whole-Cell Modeling Summer School participants.}
\begin{tabularx}{\textwidth}{l||l||X}\hline
\bfseries Group           & \bfseries Participant            & \bfseries Affiliation\\\hline\hline
Cytokinesis               & Naveen Kumar Aranganathan        & University Paris-Sud, France\\
                          & Daniel Alejandro Priego Espinosa & National Autonomous University of Mexico, Mexico\\
                          & Ilya Kiselev                     & Siberian Branch of the Russian Academy of Sciences Novosibirsk, Russia\\
                          & Wolfram Liebermeister            & Charit\'e Medical University of Berlin, Germany\\
                          & Yan Zhu                          & Monash University, Australia\\\hline
DNA repair                & Arne Bittig                      & University of Rostock, Germany\\
                          & Vijayalakshmi Chelliah           & European Bioinformatics Institute, UK\\
                          & Audald Lloret-Vilas              & European Bioinformatics Institute, UK\\
                          & Mahesh Sharma                    & National Institute of Pharmaceutical Education and Research, India\\
                          & Namrata Tomar                    & Friedrich-Alexander University of Erlangen-N\"urnberg, Germany\\\hline
Metabolism                & Kambiz Baghalian                 & University of Oxford, UK\\
                          & Frank T. Bergmann                & University of Heidelberg, Germany\\
                          & Rafael Sousa Costa               & University of Lisbon, Portugal\\
                          & Matthias K\"onig                 & Charit\'e Medical University of Berlin, Germany\\
                          & Kieran Smallbone                 & University of Manchester, UK\\
                          & Milenko Tokic                    & Swiss Federal Institute of Technology in Lausanne, Switzerland\\\hline
Protein                   & Begum Alaybeyoglu                & Bo\v{g}azi\c{c}i University, Turkey\\
                          & Matteo Cantarelli                & OpenWorm, UK\\
                          & Yin Hoon Chew                    & University of Edinburgh, UK\\
                          & Marcus Krantz                    & Humboldt University of Berlin, Germany\\
                          & Daewon Lee                       & Korea Advanced Institute of Science and Technology, Republic of Korea\\\hline
Replication               & Vincent Knight-Schrijver         & Babraham Institute, UK\\
                          & Je-Hoon Song                     & Korea Advanced Institute of Science and Technology, Republic of Korea\\
                          & Jannis Uhlendorf                 & Humboldt University of Berlin, Germany\\
                          & Dagmar Waltemath                 & University of Rostock, Germany\\
                          & James T. Yurkovich                  & University of California, San Diego, USA\\
                          & Anna Zhukova                     & National Center for Scientific Research and University of Bordeaux, France\\\hline
Replication initiation    & Harold Gomez                     & Boston University, USA\\
                          & Jens Hahn                        & Humboldt University of Berlin, Germany\\
                          & Michael Hucka                    & California Institute of Technology, USA\\
                          & Nikita Mandrik                   & Siberian Branch of the Russian Academy of Sciences Novosibirsk, Russia\\
                          & Martin Scharm                    & University of Rostock, Germany\\
                          & Florian Wendland                 & University of Rostock, Germany\\\hline
RNA                       & Tuure Hameri                     & Swiss Federal Institute of Technology in Lausanne, Switzerland\\
                          & J. Kyle Medley                   & University of Washington, USA\\
                          & Sucheendra Kumar Palaniappan     & Institute for Research in Computer Science and Automation, France\\
                          & Pinar Pir                        & Babraham Institute, UK\\
                          & Natalie Stanford                 & University of Manchester, UK\\
                          & Markus Wolfien                   & University of Rostock, Germany\\\hline
Translation               & Joseph Cursons                   & University of Melbourne, Australia\\
                          & Muhammad Haseeb                  & Mohammad Ali Jinnah University, Pakistan\\
                          & Daniel Hernandez                 & Swiss Federal Institute of Technology in Lausanne, Switzerland\\
                          & Denis Kazakiewicz                & University of Hasselt, Belgium and Medical University of Białystok, Poland\\
                          & Pedro Mendes                     & University of Manchester, UK\\
                          & Hojjat Naderi Meshkin            & Academic Center for Education, Culture and Research, Iran\\\hline
Integration               & Paulo Eduardo Pinto Burke        & Federal University of S\~ao Paulo, Brazil\\
                          & Tobias Czauderna                 & Monash University, Australia\\
                          & Bertrand Moreau                  & CoSMo Company, France\\
                          & Chris J. Myers                   & University of Utah, USA\\
                          & Thawfeek Mohamed Varusai         & University College Dublin, Ireland\\
                          & Argyris Zardilis                 & University of Edinburgh, UK\\\hline
Annotation, documentation & Christian Kn\"upfer              & University of Jena, Germany\\
and visualization         & Falk Schreiber                   & Monash University, Australia\\
                          & Tom Theile                       & University of Rostock, Germany\\\hline
Modeling instructor       & Jonathan R. Karr                 & Icahn School of Medicine at Mount Sinai, USA\\\hline
\end{tabularx}
\end{table*}

\end{document}
